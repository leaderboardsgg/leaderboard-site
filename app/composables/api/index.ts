export { default as useChangePassword } from './useChangePassword'
export { default as useConfirmAccount } from './useConfirmAccount'
export { default as useGetLeaderboardBySlug } from './useGetLeaderboardBySlug'
export { default as useGetLeaderboards } from './useGetLeaderboards'
export { default as useGetUser } from './useGetUser'
export { default as useLoginUser } from './useLoginUser'
export { default as useLogoutUser } from './useLogoutUser'
export { default as useRecoverAccount } from './useRecoverAccount'
export { default as useRegisterUser } from './useRegisterUser'
export { default as useValidateRecoveryToken } from './useValidateRecoveryToken'
export { default as useResendAccountConfirmation } from './useResendAccountConfirmation'
