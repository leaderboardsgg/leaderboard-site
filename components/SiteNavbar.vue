<template>
  <div class="border-b border-gray-300">
    <div
      class="
        container relative
        flex flex-wrap lg:flex-row justify-between
        mx-auto px-3 py-3
      "
    >
      <NuxtLink class="flex flex-grow" to="/">
        <div class="flex">
          <svg
            xmlns="http://www.w3.org/2000/svg"
            class="h-10 w-10 mx-1"
            fill="none"
            viewBox="0 0 24 24"
            stroke="currentColor"
          >
            <path
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              d="M12 8v4l3 3m6-3a9 9 0 11-18 0 9 9 0 0118 0z"
            />
          </svg>
          <h1
            class="
              w-full
              py-2
              text-lg text-center md:text-left
              font-semibold
              leading-6
              tracking-tight
              hidden md:inline
            "
          >
            speedrun.website
          </h1>
        </div>
      </NuxtLink>
      <div
        v-show="lg || mobileNavIsActive"
        class="
          border-t border-gray-300 lg:border-t-0
          flex flex-col lg:flex-row
          mt-2 lg:mt-0
          w-full lg:w-auto
          space-y-2 lg:space-y-0
        "
      >
        <CoreNavLinks />
        <CoreSearchBar />
      </div>
      <div class="
             flex flex-row items-center
             absolute right-2 top-4
             lg:relative lg:right-0 lg:top-0
             space-x-2 lg:space-x-0
           "
      >
        <CoreLoginButton />
        <CoreSignUpButton />
        <button
          class="
            items-center
            rounded
            text-black
            lg:hidden
          "
          @click="mobileNavIsActive = !mobileNavIsActive"
        >
          <svg
            class="fill-current h-5 w-5 mx-2"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>Menu</title>
            <path d="M0 3h20v2H0V3zm0 6h20v2H0V9zm0 6h20v2H0v-2z" />
          </svg>
        </button>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'

/**
 * TODO use https://vueuse.org/core/usebreakpoints and
 * https://vueuse.org/core/useEventListener/
 */

export default Vue.extend({
  data: () => ({
    lg: false,
    mobileNavIsActive: false,
  }),
  mounted() {
    this.lgSetBool()
  },
  beforeMount() {
    addEventListener('resize', this.lgSetBool)
  },
  beforeDestroy() {
    removeEventListener('resize', this.lgSetBool)
  },
  methods: {
    lgSetBool() {
      return (this.lg = innerWidth >= 1024)
    },
  },
})
</script>
