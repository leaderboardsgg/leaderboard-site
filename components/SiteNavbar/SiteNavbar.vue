<template>
  <div class="border-b border-gray-300">
    <div
      class="
        container
        relative
        flex flex-wrap
        lg:flex-row
        justify-between
        mx-auto
        px-3
        py-3
      "
    >
      <div class="flex flex-grow">
        <NuxtLink class="flex" to="/">
          <svg
            xmlns="http://www.w3.org/2000/svg"
            class="h-10 w-10 mx-1"
            fill="none"
            viewBox="0 0 24 24"
            stroke="currentColor"
          >
            <path
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              d="M12 8v4l3 3m6-3a9 9 0 11-18 0 9 9 0 0118 0z"
            />
          </svg>
          <h1
            class="
              w-full
              py-2
              text-lg text-center
              md:text-left
              font-semibold
              leading-6
              tracking-tight
              hidden
              md:inline
            "
          >
            leaderboards.gg
          </h1>
        </NuxtLink>
      </div>
      <div
        :class="{ hidden: !mobileNavIsActive }"
        class="
          lg:flex
          border-t border-gray-300
          lg:border-t-0
          flex-col
          lg:flex-row
          mt-2
          lg:mt-0
          w-full
          lg:w-auto
          space-y-2
          lg:space-y-0
        "
      >
        <CoreNavLinks />
        <CoreSearchBar />
      </div>
      <div
        class="
          flex flex-row
          items-center
          absolute
          right-2
          top-4
          lg:relative lg:right-0 lg:top-0
          space-x-2
          lg:space-x-0
        "
      >
        <CoreLoginButton
          data-testId="site-navbar-login-button"
          @click="toggleLoginModal"
          @keyup.enter="toggleLoginModal"
        />
        <CoreSignUpButton
          data-testId="site-navbar-sign-up-button"
          @click="toggleSignUpModal"
          @keyup.enter="toggleSignUpModal"
        />

        <button
          class="items-center rounded text-black lg:hidden"
          @click="toggleMenu"
        >
          <svg
            class="fill-current h-5 w-5 mx-2"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>Menu</title>
            <path d="M0 3h20v2H0V3zm0 6h20v2H0V9zm0 6h20v2H0v-2z" />
          </svg>
        </button>
      </div>
    </div>

    <transition
      enter-active-class="transition-opacity duration-200"
      leave-active-class="transition-opacity duration-200"
      enter-class="opacity-0"
      enter-to-class="opacity-100"
      leave-class="opacity-100"
      leave-to-class="opacity-0"
    >
      <BaseModal v-show="showModal" @close="showModal = false">
        <LogInCard
          v-show="showLogin"
          class="shadow-xl"
          :modal="true"
          @close="showModal = false"
          @signUpClick="showLogin = false"
        />
        <SignUpCard
          v-show="!showLogin"
          class="shadow-xl"
          :modal="true"
          @close="showModal = false"
          @logInClick="showLogin = true"
        />
      </BaseModal>
    </transition>
  </div>
</template>

<script lang="ts">
import { defineComponent, reactive, toRefs } from '@nuxtjs/composition-api';
import LogInCard from '../LogInCard/LogInCard.vue';
import SignUpCard from '../SignUpCard/SignUpCard.vue';

export default defineComponent({
  components: {
    LogInCard,
    SignUpCard,
  },
  setup() {
    const state = reactive({
      mobileNavIsActive: false,
      showLogin: false,
      showModal: false,
    });

    function toggleMenu() {
      state.mobileNavIsActive = !state.mobileNavIsActive;
    }

    function toggleLoginModal() {
      state.showModal = !state.showModal;
      state.showLogin = true;
    }

    function toggleSignUpModal() {
      state.showModal = !state.showModal;
      state.showLogin = false;
    }

    return {
      ...toRefs(state),
      toggleLoginModal,
      toggleMenu,
      toggleSignUpModal,
    };
  },
});
</script>
