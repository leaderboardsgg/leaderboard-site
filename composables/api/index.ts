export { default as useLoginUser } from './useLoginUser'
export { default as useLogoutUser } from './useLogoutUser'
export { default as useRegisterUser } from './useRegisterUser'
