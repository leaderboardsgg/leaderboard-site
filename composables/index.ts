export { default as useCurrentUser } from './useCurrentUser'
export { default as useSessionToken } from './useSessionToken'
