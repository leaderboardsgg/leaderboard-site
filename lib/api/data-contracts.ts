/* eslint-disable */
/* tslint:disable */
/*
 * ---------------------------------------------------------------
 * ## THIS FILE WAS GENERATED VIA SWAGGER-TYPESCRIPT-API        ##
 * ##                                                           ##
 * ## AUTHOR: acacode                                           ##
 * ## SOURCE: https://github.com/acacode/swagger-typescript-api ##
 * ---------------------------------------------------------------
 */

export interface Ban {
  /**
   * Generated on creation.
   * @format int64
   */
  id?: number

  /** Can't be <code>null</code>. */
  reason: string

  /**
   * Generated on creation.
   * @format date-time
   */
  time: string

  /**
   * ID of User who set the Ban. Must be either an admin or a mod.
   * @format uuid
   */
  banningUserId?: string | null

  /**
   * ID of User who received the Ban.
   * @format uuid
   */
  bannedUserId: string

  /**
   * ID of Leaderboard this Ban belongs to. If <code>null</code>, then this Ban is site-wide.
   * @format int64
   */
  leaderboardId?: number | null
}

/**
 * A Category tied to a Leaderboard.
 */
export interface Category {
  /**
   * The Category's ID. Generated on creation.
   * @format int64
   */
  id?: number

  /**
   * The Category's name.
   * @example Mongolian Throat Singing%
   */
  name: string

  /**
   * The Category's slug. <br />
   * Must be 2-25 characters inclusive and only consist of letters, numbers, and
   * hyphens.
   * @example mongolian-throat-singing
   */
  slug: string

  /** Category-specific rules. */
  rules?: string | null

  /**
   * Minimum player count for this Category. Defaults to 1.
   * @format int32
   */
  playersMin: number

  /**
   * Maximum player count for this Category. Defaults to PlayersMin.
   * @format int32
   */
  playersMax: number

  /**
   * ID of the Leaderboard this Category belongs to.
   * @format int64
   */
  leaderboardId: number
}

/**
 * Request object sent when creating a Category.
 */
export interface CreateCategoryRequest {
  /**
   * Name for the new Category.
   * @example Mongolian Throat Singing%
   */
  name: string

  /**
   * The bit in the URL that uniquely identifies this Category. <br />
   * E.g.: https://leaderboards.gg/slug-for-board/slug-for-category <br />
   * Must be 2-25 characters inclusive and only consist of letters, numbers, and
   * hyphens.
   * @example mongolian-throat-singing
   */
  slug: string

  /** Category-specific rules. */
  rules?: string | null

  /**
   * Minimum player count for this Category. Defaults to 1.
   * @format int32
   * @min 1
   * @max 2147483647
   */
  playersMin?: number | null

  /**
   * Maximum player count for this Category. Defaults to PlayersMin.
   * @format int32
   * @min 1
   * @max 2147483647
   */
  playersMax?: number | null

  /**
   * ID of the Leaderboard this Category belongs to.
   * @format int64
   */
  leaderboardId: number
}

/**
 * Request object sent when creating a Judgement.
 */
export interface CreateJudgementRequest {
  /**
   * GUID of the run.
   * @format uuid
   * @example e1c010d7-b499-4196-be17-aa27a053f3dc
   */
  runId?: string

  /**
   * Judgement comments. Must be provided if not outright approving a run ("Approved" is false or null).
   * Acts as mod feedback for the runner.
   * @example e1c010d7-b499-4196-be17-aa27a053f3dc
   */
  note?: string | null

  /** The judgement result. Can be true, false, or null. For the latter two, "Note" must be non-empty. */
  approved?: boolean | null
}

export interface CreateLeaderboardBanRequest {
  /**
   * The ID of the user, who should be banned.
   * @format uuid
   */
  userId: string

  /** The reason why the user is banned. */
  reason: string

  /**
   * The ID of the leaderboard on which the user should be banned.
   * @format int64
   */
  leaderboardId?: number
}

export interface CreateLeaderboardRequest {
  name?: string | null
  slug?: string | null
}

/**
 * Request object sent when setting a User as mod for a Leaderboard.
 */
export interface CreateModshipRequest {
  /**
   * The Leaderboard ID.
   * @format int64
   */
  leaderboardId: number

  /**
   * The User ID.
   * @format uuid
   */
  userId: string
}

export interface CreateParticipationRequest {
  comment?: string | null
  vod?: string | null

  /** @format uuid */
  runnerId: string

  /** @format uuid */
  runId: string
  isSubmitter: boolean
}

export interface CreateRunRequest {
  /** @format date-time */
  played: string

  /** @format date-time */
  submitted: string

  /**
   * 0: Created
   * 1: Submitted
   * 2: Pending
   * 3: Approved
   * 4: Rejected
   */
  status: RunStatus
}

export interface CreateSiteBanRequest {
  /**
   * The ID of the user, who should be banned.
   * @format uuid
   */
  userId: string

  /** The reason why the user is banned. */
  reason: string
}

/**
 * A decision by a mod on a run submission.
 */
export interface Judgement {
  /**
   * Generated on creation.
   * @format int64
   */
  id?: number

  /**
   * Defines this judgement, which in turn defines the status of its related run. <br />
   * If:
   *   <ul><li>true, run is approved;</li><li>false, run is rejected;</li><li>null, run is commented on.</li></ul>
   * For the latter two, Note MUST be non-empty.
   */
  approved?: boolean | null

  /**
   * When the judgement was made.
   * @format date-time
   */
  createdAt: string

  /**
   * Comments on the judgement.
   * MUST be non-empty for rejections or comments (Approved âˆˆ {false, null}).
   */
  note: string

  /**
   * ID of the mod that made this judgement.
   * @format uuid
   */
  modId: string
  mod: User

  /**
   * ID of the related run.
   * @format uuid
   */
  runId: string
  run: Run
}

/**
 * A decision by a mod on a run submission. See Models/Entities/Judgement.cs.
 */
export interface JudgementViewModel {
  /**
   * The newly-made judgement's ID.
   * @format int64
   */
  id?: number

  /** The judgement result. Can be true, false, or null. In the latter two, <code>Note</code> will be non-empty. */
  approved?: boolean | null

  /**
   * When the judgement was made. Follows <a href="https://www.ietf.org/rfc/rfc3339.txt">RFC 3339</a>.
   * @example 2022-01-01T12:34:56Z / 2022-01-01T12:34:56+01:00
   */
  createdAt?: string | null

  /**
   * Judgement comments. Acts as mod feedback for the runner. Will be non-empty for
   * non-approval judgements (Approved is false or null).
   */
  note?: string | null

  /**
   * ID of mod who made this judgement.
   * @format uuid
   */
  modId?: string

  /**
   * ID of run this judgement's for.
   * @format uuid
   */
  runId?: string
}

export interface Leaderboard {
  /**
   * Generated on creation.
   * @format int64
   */
  id?: number

  /**
   * The Leaderboard's aka game's name. Pretty straightforward.
   * @example Mario Goes to Jail II
   */
  name: string

  /**
   * The bit in the URL after the domain that can be used to identify a Leaderboard.
   * Meant to be human-readable. It must be:
   * <ul><li>between 2-80 characters, inclusive</li><li>a string of characters separated by hyphens, if desired</li></ul>
   * @example mario-goes-to-jail-ii
   */
  slug: string

  /**
   * The general rules for the Leaderboard.<br />
   * Category-specific rules are tied to the Category.
   * @example Timer starts on selecting New Game and ends when the first tear drops.
   */
  rules?: string | null
}

/**
 * Request object sent when logging a User in.
 */
export interface LoginRequest {
  /**
   * User's email.
   * @format email
   * @example ayylmao.gaming@alg.gg
   */
  email: string

  /**
   * User's password. It:
   * <ul><li>must be 8-80 characters long, inclusive;</li><li>must have at least:</li><ul><li>an uppercase letter;</li><li>a lowercase letter; and</li><li>a number.</li></ul><li>supports Unicode.</li></ul>
   * @example P4ssword
   */
  password: string
}

/**
 * Response object received on a successful login.
 */
export interface LoginResponse {
  /** A JWT to authenticate and authorize future queries with. */
  token: string
}

export interface Modship {
  /**
   * Generated on creation.
   * @format int64
   */
  id?: number

  /**
   * The mod's ID.
   * @format uuid
   */
  userId: string
  user?: User

  /**
   * ID of the Leaderboard the User is a mod for.
   * @format int64
   */
  leaderboardId: number
  leaderboard?: Leaderboard
}

export interface Participation {
  /** @format int64 */
  id?: number
  comment?: string | null
  vod?: string | null

  /** @format uuid */
  runnerId: string
  runner: User

  /** @format uuid */
  runId: string
  run: Run
}

export interface ProblemDetails {
  type?: string | null
  title?: string | null

  /** @format int32 */
  status?: number | null
  detail?: string | null
  instance?: string | null
}

/**
 * Request object sent when registering a User.
 */
export interface RegisterRequest {
  /**
   * The username to register with. It must be:
   *   <ul><li>2-25 characters long, inclusive;</li><li>made up of letters sandwiching zero or one of:</li><ul><li>hyphen;</li><li>underscore; or</li><li>apostrophe</li></ul></ul>
   * Usernames are also saved with casing, but matched without. This means you won't
   * be able to register as "Cool" if someone already called "cool" exists.
   * @pattern (?:[a-zA-Z0-9][-_']?){1,12}[a-zA-Z0-9]
   * @example Ayy-l'maoGaming
   */
  username: string

  /**
   * User's email.
   * @format email
   * @example ayylmao.gaming@alg.gg
   */
  email: string

  /**
   * User's password. It:
   * <ul><li>must be 8-80 characters long, inclusive;</li><li>must have at least:</li><ul><li>an uppercase letter;</li><li>a lowercase letter; and</li><li>a number.</li></ul><li>supports Unicode.</li></ul>
   * @example P4ssword
   */
  password: string

  /**
   * Password confirmation. It must match <code>password</code>.
   * @example P4ssword
   */
  passwordConfirm: string
}

/**
 * Request object sent when removing a User as mod for a Leaderboard.
 */
export interface RemoveModshipRequest {
  /**
   * The Leaderboard ID.
   * @format int64
   */
  leaderboardId: number

  /**
   * The User ID.
   * @format uuid
   */
  userId: string
}

export interface Run {
  /** @format uuid */
  id?: string

  /** @format date-time */
  played: string

  /** @format date-time */
  submitted: string

  /**
   * 0: Created
   * 1: Submitted
   * 2: Pending
   * 3: Approved
   * 4: Rejected
   */
  status: RunStatus
  judgements?: Judgement[] | null
  participations: Participation[]
}

/**
* 0: Created
1: Submitted
2: Pending
3: Approved
4: Rejected
* @format int32
*/
export type RunStatus = 0 | 1 | 2 | 3 | 4

export interface UpdateParticipationRequest {
  comment?: string | null
  vod: string
}

export interface User {
  /**
   * A GUID that identifies the User. Generated on creation.
   * @format uuid
   * @example 4b3835ca-dee1-4019-82b4-d2d26a7cce74
   */
  id?: string

  /**
   * The User's name. Must be:
   * <ul><li>between 2 - 25 characters inclusive; and</li><li>an alphanumeric sequence, each separated by zero or one of:</li><ul><li>an underscore;</li><li>a hyphen; or</li><li>an apostrophe</li></ul></ul>
   * Saving a name is case-sensitive, but matching against existing Users won't be.
   * @example Ayylmao Gaming
   */
  username: string

  /**
   * The User's email. Must be, well, an email.
   * @example ayylmao.gaming@alg.gg
   */
  email: string

  /** User's about text. I.e. a personal description. */
  about?: string | null

  /** User's admin status. */
  admin: boolean
  bansGiven?: Ban[] | null
  bansReceived?: Ban[] | null
  modships?: Modship[] | null
  participations?: Participation[] | null
}

export interface LeaderboardsListParams {
  /** The IDs. */
  ids?: number[]
}
