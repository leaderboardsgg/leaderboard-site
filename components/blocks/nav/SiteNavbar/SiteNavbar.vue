<script setup lang="ts">
import { computed, reactive } from 'vue'
import NavLinks from 'elements/nav/NavLinks/NavLinks.vue'
import LogoutButton from 'elements/buttons/LogoutButton/LogoutButton.vue'
import SignUpButton from 'elements/buttons/SignUpButton/SignUpButton.vue'
import LoginButton from 'elements/buttons/LoginButton/LoginButton.vue'
import LogInCard from 'blocks/cards/LogInCard/LogInCard.vue'
import SignUpCard from 'blocks/cards/SignUpCard/SignUpCard.vue'
import BaseModal from 'elements/modals/BaseModal/BaseModal.vue'
import SearchBar from 'blocks/SearchBar/SearchBar.vue'
import { useLogoutUser } from 'composables/api'
import { User } from 'lib/api/data-contracts'

interface NavbarState {
  mobileNavIsActive: boolean
  showLogin: boolean
  showModal: boolean
}

const state: NavbarState = reactive({
  mobileNavIsActive: false,
  showLogin: false,
  showModal: false,
})

const currentUser = useState<User>('current_user', () => ({
  admin: false,
  email: '',
  username: '',
}))
const loggedIn = computed(
  () => !!currentUser.value?.username && currentUser.value?.username !== '',
)

function toggleMenu() {
  state.mobileNavIsActive = !state.mobileNavIsActive
}

function toggleLoginModal() {
  state.showModal = !state.showModal
  state.showLogin = true
}

function toggleSignUpModal() {
  state.showModal = !state.showModal
  state.showLogin = false
}

function logout() {
  useLogoutUser()
}
</script>

<template>
  <div class="site-navbar">
    <div class="site-navbar__wrapper">
      <div class="site-logo">
        <NuxtLink class="site-navbar__homelink" to="/">
          <i-svg-clock class="mx-1 h-10 w-10" />
          <h1>leaderboards.gg</h1>
        </NuxtLink>
      </div>
      <div :class="{ hidden: !state.mobileNavIsActive }" class="mobile-menu">
        <NavLinks />
        <SearchBar />
      </div>
      <div class="mobile-navbar">
        <LoginButton
          v-if="!loggedIn"
          data-testId="site-navbar-login-button"
          @click="toggleLoginModal"
          @keyup.enter="toggleLoginModal"
        />
        <SignUpButton
          v-if="!loggedIn"
          data-testId="site-navbar-sign-up-button"
          @click="toggleSignUpModal"
          @keyup.enter="toggleSignUpModal"
        />
        <LogoutButton
          v-if="loggedIn"
          data-testId="site-navbar-logout-button"
          @click="logout"
          @keyup.enter="logout"
        />

        <button class="mobile-menu__toggle" @click="toggleMenu">
          <i-svg-menu class="mx-2 h-5 w-5 fill-current" />
        </button>
      </div>
    </div>

    <transition
      v-if="!loggedIn"
      enter-active-class="transition-opacity duration-200"
      leave-active-class="transition-opacity duration-200"
      enter-class="opacity-0"
      enter-to-class="opacity-100"
      leave-class="opacity-100"
      leave-to-class="opacity-0"
    >
      <BaseModal v-show="state.showModal" @close="state.showModal = false">
        <LogInCard
          v-show="state.showLogin"
          class="shadow-xl"
          :modal="true"
          @close="state.showModal = false"
          @sign-up-click="state.showLogin = false"
        />
        <SignUpCard
          v-show="!state.showLogin"
          class="shadow-xl"
          :modal="true"
          @close="state.showModal = false"
          @log-in-click="state.showLogin = true"
        />
      </BaseModal>
    </transition>
  </div>
</template>

<style lang="postcss" scoped>
.site-navbar {
  @apply border-b border-gray-300;
  & .site-navbar__wrapper {
    @apply container relative flex flex-wrap lg:flex-row justify-between mx-auto px-3 py-3;
    & .mobile-menu {
      @apply lg:flex border-t border-gray-300 lg:border-t-0 flex-col lg:flex-row mt-2 lg:mt-0 w-full lg:w-auto space-y-2 lg:space-y-0;
    }
    & .mobile-navbar {
      @apply flex flex-row items-center absolute right-2 top-4 lg:relative lg:right-0 lg:top-0 space-x-2 lg:space-x-0;
      & .mobile-menu__toggle {
        @apply items-center rounded text-black lg:hidden;
      }
    }
  }
  & .site-logo {
    @apply flex grow;
    & .site-navbar__homelink {
      @apply flex;
      & h1 {
        @apply w-full py-2 text-lg text-center md:text-left font-semibold leading-6 tracking-tight hidden md:inline;
      }
    }
  }
}
</style>
