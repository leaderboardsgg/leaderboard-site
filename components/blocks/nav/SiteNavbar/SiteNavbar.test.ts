// import { fileURLToPath } from 'node:url'
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-expect-error
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { setup, $fetch, url, startServer, setup } from '@nuxt/test-utils'
import { describe, expect, test, vi, beforeEach, afterEach } from 'vitest'

// import from 'nuxt'

import { ref } from 'vue'
import SiteNavbar from './SiteNavbar.vue'
// import 'root/vitest.setup'
import { fireEvent, stubbedRender } from 'root/testUtils'
import * as apiComposables from 'composables/api'
import { User } from 'lib/api/data-contracts'

await setup({
  rootDir: './../../../../../',
  server: true,
})

// vi.doMock(url('#app'), () => ({
//   useRuntimeConfig: () => ({
//     public: {
//       BACKEND_BASE_URL: process.env.BACKEND_BASE_URL,
//     },
//   }),
// }))

afterEach(() => {
  vi.restoreAllMocks()
  vi.clearAllMocks()
})

describe('<SiteNavbar />', () => {
  test('should render without crashing', () => {
    const { unmount } = stubbedRender(SiteNavbar)

    unmount()
  })

  describe('when no user is logged in', () => {
    beforeEach(() => {
      ref<User>({
        admin: false,
        email: '',
        username: '',
      })
    })

    test('should render the login/sign up buttons', () => {
      const { getByTestId } = stubbedRender(SiteNavbar)

      expect(getByTestId('site-navbar-login-button')).toBeInTheDocument()
      expect(getByTestId('site-navbar-sign-up-button')).toBeInTheDocument()
    })

    describe('when the login button is clicked', () => {
      test('should bring up the login card', async () => {
        const { container, getByTestId } = stubbedRender(SiteNavbar)

        await fireEvent.click(getByTestId('site-navbar-login-button'))

        expect(container.querySelector('.login-card')).toBeInTheDocument()
      })
    })

    describe('when the sign up button is clicked', () => {
      test('should bring up the signup card', async () => {
        const { container, getByTestId } = stubbedRender(SiteNavbar)

        await fireEvent.click(getByTestId('site-navbar-sign-up-button'))

        expect(container.querySelector('.signup-card')).toBeInTheDocument()
      })
    })
  })

  describe('when a user is logged in', () => {
    beforeEach(() => {
      ref<User>({
        admin: true,
        email: 'admin@leaderboards.gg',
        username: 'lbgg_admin',
      })
    })

    test('should render the logout button', () => {
      const { getByTestId } = stubbedRender(SiteNavbar)

      expect(getByTestId('site-navbar-logout-button')).toBeInTheDocument()
    })

    describe('when the logout button is clicked', () => {
      const useLogoutUserSpy = vi.spyOn(apiComposables, 'useLogoutUser')

      test('should log out the user', async () => {
        const { getByTestId } = stubbedRender(SiteNavbar)

        await fireEvent.click(getByTestId('site-navbar-logout-button'))

        expect(useLogoutUserSpy).toHaveBeenCalled()
      })
    })
  })
})
