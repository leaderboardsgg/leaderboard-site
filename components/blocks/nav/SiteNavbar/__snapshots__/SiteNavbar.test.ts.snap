// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SiteNavbar /> renders correctly 1`] = `
<div
  class="site-navbar"
>
  <div
    class="site-navbar__wrapper"
  >
    <div
      class="site-logo"
    >
      <a
        class="site-navbar__homelink"
        href="/"
      >
        <svg-icon-stub
          class="mx-1 w-10 h-10"
          name="clock"
        />
         
        <h1>
          leaderboards.gg
        </h1>
      </a>
    </div>
     
    <div
      class="mobile-menu hidden"
    >
      <nav
        class="nav-link-container"
      >
        <a
          class="button-link nav-link"
          href=""
          name="Games"
          to="#"
        >
          
    Games
  
        </a>
        <a
          class="button-link nav-link"
          href=""
          name="About"
          to="#"
        >
          
    About
  
        </a>
      </nav>
       
      <div
        class="search-bar"
      >
        <input
          data-testid="search-input"
          name="search"
          placeholder="search..."
        />
         
        <button
          data-testid="search-button"
          type="button"
        >
          <svg-icon-stub
            name="search"
          />
        </button>
      </div>
    </div>
     
    <div
      class="mobile-navbar"
    >
      <button
        class="button login-button"
        data-testid="site-navbar-login-button"
      >
        
  Log In

      </button>
       
      <button
        class="button sign-up-button"
        data-testid="site-navbar-sign-up-button"
      >
        
  Sign Up

      </button>
       
      <!---->
       
      <button
        class="mobile-menu__toggle"
      >
        <svg-icon-stub
          class="mx-2 w-5 h-5 fill-current"
          name="menu"
        />
      </button>
    </div>
  </div>
   
  <transition-stub
    enteractiveclass="transition-opacity duration-200"
    enterclass="opacity-0"
    entertoclass="opacity-100"
    leaveactiveclass="transition-opacity duration-200"
    leaveclass="opacity-100"
    leavetoclass="opacity-0"
  >
    <div
      class="modal"
      style="display: none;"
    >
      <div
        class="modal__wrapper"
      >
        <div
          class="modal__container"
        >
          <div
            class="modal-fade"
          />
           
          <div
            class="modal__content"
          >
            <div
              class="card login-card shadow-xl"
              data-testid="login-card"
              id="loginCard"
              style="display: none;"
            >
              <div
                class="card-header login-card__header"
              >
                <div
                  class="login-card__title"
                >
                  Log In
                </div>
                 
                <button
                  class="button login-card__sign-up-button"
                  id="sign-up-button"
                >
                  
      Sign Up
    
                </button>
                 
                <button
                  class="button close-button"
                >
                  <svg-icon-stub
                    class="w-5 h-5"
                    name="close"
                  />
                </button>
              </div>
               
              <div
                class="card-body"
              >
                <div
                  class="login-card__body-wrapper"
                >
                  <input
                    autocomplete="email"
                    class="input"
                    name="email"
                    placeholder="Email"
                    type="text"
                  />
                   
                  <div
                    class="login-card__input-wrapper"
                  >
                    <input
                      autocomplete="password"
                      class="input login-card__password-field"
                      name="password"
                      placeholder="Password"
                      type="password"
                    />
                     
                    <div
                      class="login-card__button-wrapper"
                    >
                      <button
                        class="button button"
                        data-testid="hide-show-password-button"
                        id="hide-show-button"
                        type="button"
                      >
                        <svg-icon-stub
                          class="w-5 h-5"
                          data-testid="hidden-eye-icon"
                          name="eye-hidden"
                        />
                         
                        <svg-icon-stub
                          class="w-5 h-5"
                          data-testid="visible-eye-icon"
                          name="eye-visible"
                          style="display: none;"
                        />
                      </button>
                    </div>
                  </div>
                   
                  <button
                    class="button login-button"
                  >
                     Log In 
                  </button>
                </div>
                 
                <div
                  class="login-card__auth-buttons"
                >
                  <button
                    class="button login-button"
                  >
                    <svg-icon-stub
                      class="mr-2 w-5 h-5"
                      name="github"
                    />
                     
                    <p>
                      Log In with Github
                    </p>
                  </button>
                   
                  <button
                    class="button login-button"
                  >
                    <svg-icon-stub
                      class="mr-2 w-5 h-5"
                      name="google"
                    />
                     
                    <p>
                      Log In with Google
                    </p>
                  </button>
                </div>
              </div>
            </div>
             
            <div
              class="card signup-card shadow-xl"
              data-testid="sign-up-card"
              id="signUpCard"
            >
              <div
                class="card-header signup-card__header"
              >
                <button
                  class="button signup-card__login-button"
                  id="login-button"
                >
                  
      Log In
    
                </button>
                 
                <div
                  class="signup-card__title"
                >
                  Sign Up
                </div>
                 
                <button
                  class="button close-button"
                >
                  <svg-icon-stub
                    class="w-5 h-5"
                    name="close"
                  />
                </button>
              </div>
               
              <div
                class="signup-card__content"
              >
                <div
                  class="signup-card__content-wrapper"
                >
                  <input
                    autocomplete="email"
                    class="input"
                    name="email"
                    placeholder="Email"
                    type="text"
                  />
                   
                  <div
                    class="signup-card__input-wrapper"
                  >
                    <input
                      autocomplete="nickname"
                      class="input"
                      name="username"
                      placeholder="Username"
                      type="text"
                    />
                     
                    <p>
                      This can be changed later
                    </p>
                  </div>
                   
                  <div
                    class="signup-card__input-wrapper"
                  >
                    <div
                      class="signup-card__password-wrapper"
                    >
                      <input
                        autocomplete="password"
                        class="input signup-card__password-field"
                        name="password"
                        placeholder="Password"
                        type="password"
                      />
                       
                      <input
                        autocomplete="password"
                        class="input signup-card__password-field"
                        name="passwordConfirm"
                        placeholder="Confirm"
                        type="password"
                      />
                       
                      <button
                        class="button button"
                        data-testid="hide-show-password-button"
                        id="hide-show-password"
                        type="button"
                      >
                        <svg-icon-stub
                          class="w-5 h-5"
                          data-testid="hidden-eye-icon"
                          name="eye-hidden"
                        />
                         
                        <svg-icon-stub
                          class="w-5 h-5"
                          data-testid="visible-eye-icon"
                          name="eye-visible"
                          style="display: none;"
                        />
                      </button>
                    </div>
                     
                    <p>
                      * Must be 8 characters with a mix of letters and numbers
                    </p>
                  </div>
                   
                  <button
                    class="button signup-button"
                  >
                     Sign Up 
                  </button>
                </div>
                 
                <div
                  class="signup-card__auth-buttons"
                >
                  <button
                    class="button signup-button"
                  >
                    <svg-icon-stub
                      class="mr-2 w-5 h-5"
                      name="github"
                    />
                     
                    <p>
                      Sign Up with Github
                    </p>
                  </button>
                   
                  <button
                    class="button signup-button"
                  >
                    <svg-icon-stub
                      class="mr-2 w-5 h-5"
                      name="google"
                    />
                     
                    <p>
                      Sign Up with Google
                    </p>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </transition-stub>
</div>
`;
